### PR TITLE
Arrays, AST–model type split, and general type system churn

### PR DESCRIPTION
**Pull request by MattWindsor91**
**I recommend merging #100 first.  This PRQ is based on top of and thus includes #100.**

This fairly big PRQ implements arrays for Starling.  It also splits a lot of the type system logic previously shared by the AST and Starling core for cleanliness, and pushes through a few type system overhauls to allow lvalues to be things other than variables.

### CODE REVIEW

File: AST.fs [hunk 1]
```diff
@@ -352,6 +355,9 @@ module Pretty =
             match lit with
             | TInt -> hsep2 Nop (syntaxIdent (String ("int"))) suffix
             | TBool -> hsep2 Nop (syntaxIdent (String ("bool"))) suffix
+            | TArray (len, contents) ->
+                let lenSuffix = squared (String (sprintf "%d" len))
+                pl contents (hsep2 Nop suffix lenSuffix)
```
Comment (septract) [inline AST.fs]: use `<->` here?

File: AST.fs [hunk 2]
```diff
@@ -481,17 +487,35 @@ let (|ArithIn|BoolIn|AnyIn|) : BinOp -> Choice<unit, unit, unit> =
...
+    | BoolExp' _ -> BoolExp e
+    | ArithExp' _ -> ArithExp e
+    | AnyExp' _ -> AnyExp e
+
+/// <summary>
+///     Active pattern classifying expressions as lvalues or rvalues.
+/// </summary>
+let (|LValue|RValue|) (e : Expression) : Choice<Expression, Expression> =
+    match e.Node with
+    | Identifier _ | Symbolic _ | ArraySubscript _ -> LValue e
```
Comment (septract) [inline AST.fs]: Symbolic is an lvalue??

Comment (septract) [reply AST.fs]: Suspicious wrt framing as well...

File: Command.fs [hunk 3]
```diff
@@ -144,6 +161,9 @@ module Compose =
             Seq.fold maxOpt None <| (Seq.map (getIntIntermediate v) <| xs)
         | IDiv (x, y) | IMod (x, y) ->
             maxOpt (getIntIntermediate v x) (getIntIntermediate v y)
+        // Is this correct?
```
Comment (septract) [inline Command.fs]: should be a TODO?

File: Command.fs [hunk 4]
```diff
@@ -164,12 +184,29 @@ module Compose =
             maxOpt (getIntermediate v x) (getIntermediate v y)
         | _ -> None
 
+    /// Gets the highest intermediate number for some variable in a given
+    /// array expression
+    and getArrayIntermediate v =
```
Comment (septract) [inline Command.fs]: Type annotation

(and remove point-free if you can be bothered)

File: Expr.fs [hunk 5]
```diff
@@ -78,29 +113,40 @@ module Pretty =
...
-        >> hsep
-        >> parened
+        cmdSexpr op << Seq.map pxs
+
+    /// Pretty-prints an array subscript.
+    let rec printIdx
+      (pVar : 'Var -> Doc)
+      (arr : ArrayExpr<'Var>)
+      (idx : IntExpr<'Var>) : Doc =
+        cmdSexpr "[]" [ printIntExpr pVar idx; printArrayExpr pVar arr ]
```
Comment (septract) [inline Expr.fs]: Look up Z3 smtlib syntax for array indexes?

File: Expr.fs [hunk 6]
```diff
@@ -113,11 +159,23 @@ module Pretty =
         | BLt (x, y) -> sexpr "<" (printIntExpr pVar) [x; y]
         | BNot x -> sexpr "not" (printBoolExpr pVar) [x]
 
+    /// Pretty-prints an array expression.
+    and printArrayExpr (pVar : 'Var -> Doc) : ArrayExpr<'Var> -> Doc =
```
Comment (septract) [inline Expr.fs:163]: check against smtlib

File: Expr.fs [hunk 7]
```diff
@@ -252,6 +310,7 @@ let mkVarExp (var : CTyped<'Var>) : Expr<'Var> =
     match var with
     | Int s -> s |> IVar |> Int
     | Bool s -> s |> BVar |> Bool
+    | Array (eltype, length, s) -> Array (eltype, length, AVar s)
```
Comment (septract) [inline Expr.fs]: qualified types here?

File: GraphTests.fs [hunk 8]
```diff
@@ -387,10 +387,10 @@ module Tests =
               TestCaseData(Advisory
                                (Multiset.singleton
                                     (smgfunc BTrue "holdTick"
-                                         [ SMExpr.Int (siBefore "t") ] )))
+                                         [ Int (siBefore "t") ] )))
```
Comment (septract) [inline GraphTests.fs]: why remove the qualifier?

Comment (septract) [reply GraphTests.fs]: (all the way through this file..)

File: Model.fs [hunk 9]
```diff
@@ -52,6 +52,23 @@ open Starling.Core.Command
 /// </summary>
 [<AutoOpen>]
 module Types =
+    /// <summary>
+    ///     Mid-level register transfer logic used to encode Starling
+    ///     commands.
+    /// </summary>
+    /// <typeparam name="L">The type of lvalues.</typeparam>
+    /// <typeparam name="L">The type of rvalue variables.</typeparam>
```
Comment (septract) [inline Model.fs]: typo: name="RV"

File: Modeller.fs [hunk 10]
```diff
@@ -412,8 +414,10 @@ module Pretty =
 (*
  * Starling imperative language semantics
  *)
-let prim : string -> TypedVar list -> TypedVar list -> SVBoolExpr -> PrimSemantics =
-    fun name results args body -> { Name = name; Results = results; Args = args; Body = body }
+let prim (name : string) (results : TypedVar list) (args : TypedVar list)
```
Comment (septract) [inline Modeller.fs]: MkPrim?

File: Modeller.fs [hunk 11]
```diff
@@ -570,16 +579,29 @@ let rec modelExpr
...
+                    match eltype with
+                    | Int () -> Int (IIdx (eltype, length, arrE, idxE))
+                    | Bool () -> Bool (BIdx (eltype, length, arrE, idxE))
+                    | Array (ieltype, ilength, ()) ->
+                        Array (ieltype, ilength, AIdx (eltype, length, arrE, idxE)))
+                arrResult idxResult
         (* We can use the active patterns above to figure out whether we
          * need to treat this expression as arithmetic or Boolean.
          *)
         | _ -> match e with
```
Comment (septract) [inline Modeller.fs]: New active pattern here.

File: Modeller.fs [hunk 12]
```diff
@@ -1008,7 +1148,7 @@ let modelCFunc
              afunc.Params
              |> Seq.map (fun e ->
                              e
-                             |> modelExpr env id
+                             |> modelExpr env env id
```
Comment (septract) [inline Modeller.fs]: confusing because `env` should be thr-local.

File: Modeller.fs [hunk 13]
```diff
@@ -1044,160 +1184,292 @@ let rec modelCView (ctx : MethodContext) : View -> Result<CView, ViewError> =
...
+    (* In an integer load, the destination must be a THREAD integral lvalue;
+                           the source must be a SHARED integral lvalue;
+                           and the fetch mode is unconstrained. *)
+    let modelWithExprs dstE srcE =
+        let cmd, reads =
+            match mode with
+            | Direct -> "!ILoad", [ dstE ]
+            | Increment -> "!ILoad++", [ dstE; srcE ]
+            | Decrement -> "!ILoad--", [ dstE; srcE ]
+        command cmd (List.map Int reads) [ Int srcE ]
```
Comment (septract) [inline Modeller.fs]: `reads` -> `writes`

File: Modeller.fs [hunk 14]
```diff
@@ -1044,160 +1184,292 @@ let rec modelCView (ctx : MethodContext) : View -> Result<CView, ViewError> =
...
+        (mdl ctx.ThreadVars test)
+
+/// <summary>
+///     Gets the underlying variable of an lvalue.
+/// </summary>
+/// <param name="lv">The lvalue-candidate whose type is needed.</param>
+/// <returns>
+///     The lvalue's variable; will crash if the expression is not an lvalue.
+/// </returns>
+let rec varOfLValue (lv : Expression) : string =
```
Comment (septract) [inline Modeller.fs]: Should be Var not string

File: Modeller.fs [hunk 15]
```diff
@@ -1044,160 +1184,292 @@ let rec modelCView (ctx : MethodContext) : View -> Result<CView, ViewError> =
...
+///     None.
+/// </returns>
+let typeOfLValue (env : VarMap) (lv : Expression) : Type option =
+    let matchArray var =
+        match var with
+        | Array (eltype, _, _) -> Some eltype
+        | _ -> None
+
+    // First, we work out the type of the variable at the top of the lvalue.
+    let rec findIdxType lv matcher =
```
Comment (septract) [inline Modeller.fs]: Not just Idx?

File: Modeller.fs [hunk 16]
```diff
@@ -1044,160 +1184,292 @@ let rec modelCView (ctx : MethodContext) : View -> Result<CView, ViewError> =
...
+/// <summary>
+///     Tries to get the type of an lvalue.
+/// </summary>
+/// <param name="env">The map in which the lvalue's variable exists.</param>
+/// <param name="lv">The lvalue-candidate whose type is needed.</param>
+/// <returns>
+///     If the lvalue has a valid type, the type of that lvalue; otherwise,
+///     None.
+/// </returns>
+let typeOfLValue (env : VarMap) (lv : Expression) : Type option =
```
Comment (septract) [inline Modeller.fs:1402]: Needs more documentation

File: MuZ3Backend.fs [hunk 17]
```diff
@@ -611,6 +597,7 @@ module Translator =
         match v with
         | Int iv -> Int (IVar iv)
         | Bool bv -> Bool (BVar bv)
+        | Array (eltype, length, av) -> Array (eltype, length, AVar av)
```
Comment (septract) [inline MuZ3Backend.fs]: Why is this in muZ3?

File: Semantics.fs [hunk 18]
```diff
@@ -97,14 +103,257 @@ let frameVar ctor (par : CTyped<Var>) : SMBoolExpr =
...
+    | Indices of Map<IntExpr<Sym<Var>>, Write>
+    override this.ToString () = sprintf "%A" this
+
+/// <summary>
+///     Records a write into a write map.
+/// </summary>
+/// <param name="var">The variable being written to.</param>
+/// <param name="idxPath">
+///     The path of indexes from the variable being written to to the variable.
+///     For example, [3; x; 1+i] would represent a write to A[3][x][1+i].
```
Comment (septract) [inline Semantics.fs:129]: Is this the correct order?

File: Semantics.fs [hunk 19]
```diff
@@ -97,14 +103,257 @@ let frameVar ctor (par : CTyped<Var>) : SMBoolExpr =
...
+
+        let idxRec' =
+            match idxPathRest with
+            | [] ->
+                (* If there is no subscript, then we must be writing to this
+                   entire index, so mark it as Entire. *)
+                Entire value
+            | x::xs ->
+                match idxRec with
+                | Some (Entire _) ->
```
Comment (septract) [inline Semantics.fs]: Should be an error

File: Semantics.fs [hunk 20]
```diff
@@ -97,14 +103,257 @@ let frameVar ctor (par : CTyped<Var>) : SMBoolExpr =
...
+        let idxRec = imap.TryFind idx
+        let imapLessIdx =
+            maybe imap (fun _ -> imap.Remove idx) (imap.TryFind idx)
+
+        let idxRec' =
+            match idxPathRest with
+            | [] ->
+                (* If there is no subscript, then we must be writing to this
+                   entire index, so mark it as Entire. *)
+                Entire value
```
Comment (septract) [inline Semantics.fs]: should check there's not already a value

File: Semantics.fs [hunk 21]
```diff
@@ -97,14 +103,257 @@ let frameVar ctor (par : CTyped<Var>) : SMBoolExpr =
...
+    let varRec = map.TryFind var
+    let mapLessVar = maybe map (fun _ -> map.Remove var) (map.TryFind var)
+
+    let varRec' =
+        match idxPath with
+        | [] ->
+            (* If there is no subscript, then we must be writing to this entire
+               variable, so mark it as Entire. *)
+            Entire value
+        | (x::xs) ->
```
Comment (septract) [inline Semantics.fs:180]: same issue as above.

File: Semantics.fs [hunk 22]
```diff
@@ -97,14 +103,257 @@ let frameVar ctor (par : CTyped<Var>) : SMBoolExpr =
...
+/// <returns>A well-typed <see cref="Expr"/> capturing the subscript.</returns>
+let mkIdx (eltype : Type) (length : int option) (arr : ArrayExpr<Sym<Var>>)
+  (idx : IntExpr<Sym<Var>>)
+  : Expr<Sym<Var>> =
+    let record = (eltype, length, arr, idx)
+
+    match eltype with
+    | Int () -> Int (IIdx record)
+    | Bool () -> Bool (BIdx record)
+    | Array (eltype', length', ()) -> Array (eltype', length', AIdx record)
```
Comment (septract) [inline Semantics.fs]: qualifiers

File: Semantics.fs [hunk 23]
```diff
@@ -97,14 +103,257 @@ let frameVar ctor (par : CTyped<Var>) : SMBoolExpr =
...
+       We need to convert each write map entry into an array update or a
+       direct value. *)
+    let rec translateRhs lhs =
+        function
+        | Entire v -> ok v
+        | Indices ixmap ->
+            // TODO(CaptainHayashi): proper errors.
+            match lhs with
+            | Array (eltype, length, alhs) ->
+                let addUpdate (index, value) lhs' =
```
Comment (septract) [inline Semantics.fs]: types

File: Semantics.fs [hunk 24]
```diff
@@ -97,14 +103,257 @@ let frameVar ctor (par : CTyped<Var>) : SMBoolExpr =
...
+                    (* Need to translate any further subscripts inside value.
+                       But, to do that, we need to know what the LHS of those
+                       subscripts is! *)
+                    let vlhs = mkIdx eltype length alhs index
+                    let vrhsResult = translateRhs vlhs value
+                    lift
+                        (fun (vrhs) ->
```
Comment (septract) [inline Semantics.fs]: brackets?

File: Semantics.fs [hunk 25]
```diff
@@ -97,14 +103,257 @@ let frameVar ctor (par : CTyped<Var>) : SMBoolExpr =
...
+                            Array
```
Comment (septract) [inline Semantics.fs]: qualify

File: Semantics.fs [hunk 26]
```diff
@@ -164,33 +408,98 @@ let primParamSubFun
...
+        >> mapMessages Traversal
+    let toBefore = traverseTypedSymWithMarker Before
+    let toBeforeE =
+        mapTraversal (tliftOverExpr toBefore) >> mapMessages Traversal
+    let toBeforeB =
+        mapTraversal (tliftToBoolSrc (tliftToExprDest toBefore))
+        >> mapMessages Traversal
+
+    let rec translateInstrs is =
+        // TODO(CaptainHayashi): do two-state conversion here, not earlier.
```
Comment (septract) [inline Semantics.fs]: redundant

File: Traversal.fs [hunk 27]
```diff
@@ -315,6 +353,28 @@ let tliftOverCTyped
         function
         | Int i -> lift (pairMap id Int) (sub ctx i)
         | Bool b -> lift (pairMap id Bool) (sub ctx b)
+        | Array (eltype, length, a) ->
+            lift (pairMap id (fun a -> Array (eltype, length, a))) (sub ctx a)
+
+/// <summary>
+///     Tries to extract an <see cref="ArrayExpr"/> out of an
+///     <see cref="ArrayExpr"/>, failing if the expression is not of that type
```
Comment (septract) [inline Traversal.fs]: 'arrayExpr' -> 'Expr'

File: Traversal.fs [hunk 28]
```diff
@@ -315,6 +353,28 @@ let tliftOverCTyped
...
+/// </summary>
+let expectArray
+  (eltype : Type)
+  (length : int option)
+  (expr : Expr<'Var>)
+  : Result<ArrayExpr<'Var>, TraversalError<_>> =
+    // TODO(CaptainHayashi): proper doc comment.
+    match expr with
+    | Typed.Array (_, _, ae)
+        when typesCompatible (Type.Array (eltype, length, ())) (typeOf expr) ->
```
Comment (septract) [inline Traversal.fs:372]: Expr.Array

File: Traversal.fs [hunk 29]
```diff
@@ -449,6 +511,22 @@ let rec tLiftToBoolSrc
             |> CTyped.Bool
             |> Context.changePos id sub ctx
             |> bind (uncurry (ignoreContext expectBool))
+        | BIdx (eltype, length, arr, ix) ->
+            // TODO(CaptainHayashi): this is awful.
+            let originalType = Array (eltype, length, ())
+
+            (* Ensure the traversal doesn't change eltype or length to something
+               we're not expecting. *)
+            let assemble ((eltype', length', arr'), ix') =
```
Comment (septract) [inline Traversal.fs:520]: might be unnecessary?

Comment (septract) [reply Traversal.fs:520]: ...but don't think about it too hard! >:(

File: Traversal.fs [hunk 30]
```diff
@@ -485,17 +582,84 @@ and tLiftToIntSrc
...
+                let tResult =
+                    tchain3
+                        asv
+                        (Context.inIndex isv)
+                        esv
+                        assemble ctx ((eltype, length, arr), ix, value)
+
+                // Remove the nested result.
+                bind (uncurry (ignoreContext id)) tResult
+        lift (fun (ctx, ex) -> (ctx, (eltype, length, ex))) arrayExprResult
```
Comment (septract) [inline Traversal.fs:649]: get rid of assemble?

File: TypeSystem.fs [hunk 31]
```diff
@@ -69,13 +83,39 @@ let mapCTyped (f : 'Src -> 'Dst) : CTyped<'Src> -> CTyped<'Dst> =
...
+       record of the substitutions made. *)
+    match (x, y) with
+    (* Arrays are compatible when their element types are compatible and their
+       lengths are not both defined as contradictory values. *)
+    | (Type.Array (ex, Some _ , ()), Type.Array (ey, None   , ()))
+    | (Type.Array (ex, None   , ()), Type.Array (ey, Some _ , ()))
+    | (Type.Array (ex, None   , ()), Type.Array (ey, None   , ())) ->
+        ex = ey
+    | (Type.Array (ex, Some lx, x), Typed.Array (ey, Some ly, y)) ->
+        ex = ey && lx = ly
```
Comment (septract) [inline TypeSystem.fs]: typesCompatible ex ey ?

File: Z3.fs [hunk 32]
```diff
@@ -30,16 +31,42 @@ module Pretty =
...
+    ///     The Z3 sort corresponding to <paramref name="ty"/>.
+    /// </returns>
+    let rec typeToSort (reals : bool) (ctx : Z3.Context) (ty : Type) : Z3.Sort =
+        match ty with
+        | Type.Int () when reals -> ctx.MkRealSort () :> Z3.Sort
+        | Type.Int () -> ctx.MkIntSort () :> Z3.Sort
+        | Type.Bool () -> ctx.MkBoolSort () :> Z3.Sort
+        | Type.Array (eltype, _, ()) ->
+            let elZ3 = typeToSort reals ctx eltype
+            ctx.MkArraySort (domain = elZ3, range = ctx.MkIntSort ()) :> Z3.Sort
```
Comment (septract) [inline Z3.fs]: Should do something about lengths?

File: Z3.fs [hunk 33]
```diff
@@ -53,37 +80,72 @@ module Expr =
...
         bz bool
 
+    /// Converts a Starling array expression to a Z3 Expr.
+    and arrayToZ3
+      (reals : bool)
+      (toStr : 'Var -> string)
+      (ctx: Z3.Context)
+      (eltype : Type)
+      (arr : ArrayExpr<'Var>)
+      : Z3.ArrayExpr =
```
Comment (septract) [inline Z3.fs:118]: Bug here that should be cherry-picked from mw-pldi-experimental

File: Z3.fs (hunk 32 above)
Comment (septract) [reply Z3.fs]: also bug here. 🐛 🐛 🐛

**Comment by septract** [comment]
modulo horrible bug, this is good to go :shipit:

**Comment by septract** [comment]
can I merge this?

File: Expr.fs (hunk 5 above)
Comment (MattWindsor91) [reply Expr.fs]: It's `select` (and update is `store`).  Should I change these?

File: Semantics.fs (hunk 18 above)
Comment (MattWindsor91) [reply Semantics.fs:129]: Having looked at it I believe it is

**Pull request merged**